lib/ukstore/store.c

```diff
@@ -29,3 +29,1186 @@
+	GETCASE_DOWNCASTUS(e, u64, S32, out, s32);
+	GETCASE_DOWNCASTSS(e, s64, S32, out, s32);
+	GETCASE_DOWNCASTUS(e, uptr, S32, out, s32);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s16 sscanf_ret;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+
+		sscanf_ret = sscanf(val, "%" __SCNs32, out);
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_u64(const struct uk_store_entry *e, __u64 *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, U64, out, u64);
+	GETCASE_UPCASTSU(e, s8, U64, out, u64);
+	GETCASE_UPCAST(e, u16, U64, out, u64);
+	GETCASE_UPCASTSU(e, s16, U64, out, u64);
+	GETCASE_UPCAST(e, u32, U64, out, u64);
+	GETCASE_UPCASTSU(e, s32, U64, out, u64);
+	GETCASE_UPCAST(e, u64, U64, out, u64);
+	GETCASE_UPCASTSU(e, s64, U64, out, u64);
+	GETCASE_UPCAST(e, uptr, U64, out, u64);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s16 sscanf_ret;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+
+		sscanf_ret = sscanf(val, "%" __SCNu64, out);
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_s64(const struct uk_store_entry *e, __s64 *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, S64, out, s64);
+	GETCASE_UPCAST(e, s8, S64, out, s64);
+	GETCASE_UPCAST(e, u16, S64, out, s64);
+	GETCASE_UPCAST(e, s16, S64, out, s64);
+	GETCASE_UPCAST(e, u32, S64, out, s64);
+	GETCASE_UPCAST(e, s32, S64, out, s64);
+	GETCASE_DOWNCASTUS(e, u64, S64, out, s64);
+	GETCASE_UPCAST(e, s64, S64, out, s64);
+	GETCASE_DOWNCASTUS(e, uptr, S64, out, s64);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s16 sscanf_ret;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+
+		sscanf_ret = sscanf(val, "%" __SCNs64, out);
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_uptr(const struct uk_store_entry *e, __uptr *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s8, PTR, out, uptr);
+	GETCASE_UPCAST(e, u16, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s16, PTR, out, uptr);
+	GETCASE_UPCAST(e, u32, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s32, PTR, out, uptr);
+	GETCASE_UPCAST(e, u64, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s64, PTR, out, uptr);
+	GETCASE_UPCAST(e, uptr, PTR, out, uptr);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s16 sscanf_ret;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+
+		sscanf_ret = sscanf(val, "%p", (void **) out);
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_charp(const struct uk_store_entry *e, char **out)
```

Comment:
Hum... I would actually treat the ncharp as an additional variant and keep what you have. maxlen wouldn't be usable for the other outtypes. The thing is also to define maxlen for getters that return charp already. Because our interface does imply any size limitations for charp it is likely that the output will be truncated with the ncharp varaint.